File: Sources/QRCodeSDK/Security/SecurityManager.swift
import Foundation
import CryptoKit

/// Security manager for QR code operations
public class SecurityManager {
    
    // MARK: - Rate Limiting
    
    private static var operationCounts: [String: (count: Int, lastReset: Date)] = [:]
    private static let maxOperationsPerMinute = 60
    private static let rateLimitWindow: TimeInterval = 60 // 1 minute
    private static let rateLimitQueue = DispatchQueue(label: "com.qrcodesdk.ratelimit")
    
    /// Check if operation is within rate limits (thread-safe)
    public static func checkRateLimit(for operation: String) -> Bool {
        return rateLimitQueue.sync {
            let now = Date()
            let key = operation
            
            if let existing = operationCounts[key] {
                // Reset counter if window has passed
                if now.timeIntervalSince(existing.lastReset) > rateLimitWindow {
                    operationCounts[key] = (count: 1, lastReset: now)
                    return true
                }
                
                // Check if within limits
                if existing.count >= maxOperationsPerMinute {
                    return false
                }
                
                // Increment counter
                operationCounts[key] = (count: existing.count + 1, lastReset: existing.lastReset)
                return true
            } else {
                // First operation
                operationCounts[key] = (count: 1, lastReset: now)
                return true
            }
        }
    }
    
    // MARK: - Input Sanitization
    
    /// Sanitize and validate QR code input data
    public static func sanitizeQRInput(_ input: String) throws -> String {
        // Remove null bytes and control characters
        let sanitized = input.replacingOccurrences(of: "\0", with: "")
            .components(separatedBy: .controlCharacters)
            .joined()
        
        // Validate length
        guard sanitized.count <= 4296 else { // QR Code max capacity
            throw SecurityError.inputTooLong
        }
        
        // Check for malicious patterns
        try validateInputSafety(sanitized)
        
        return sanitized
    }
    
    /// Validate input for potential security threats
    private static func validateInputSafety(_ input: String) throws {
        // Check for script injection patterns
        let dangerousPatterns = [
            "<script", "javascript:", "data:", "vbscript:",
            "onload=", "onerror=", "onclick=", "eval(",
            "document.cookie", "window.location"
        ]
        
        let lowercaseInput = input.lowercased()
        for pattern in dangerousPatterns {
            if lowercaseInput.contains(pattern) {
                throw SecurityError.potentialInjection
            }
        }
        
        // Validate URL schemes if input looks like URL
        if input.hasPrefix("http") || input.hasPrefix("ftp") {
            try validateURLSafety(input)
        }
    }
    
    /// Validate URL safety for QR codes containing URLs
    private static func validateURLSafety(_ urlString: String) throws {
        guard let url = URL(string: urlString) else {
            throw SecurityError.invalidURL
        }
        
        // Whitelist allowed schemes
        let allowedSchemes = ["https", "http", "tel", "mailto", "sms"]
        guard let scheme = url.scheme?.lowercased(),
              allowedSchemes.contains(scheme) else {
            throw SecurityError.unsafeURLScheme
        }
        
        // Check for suspicious domains
        if let host = url.host?.lowercased() {
            let suspiciousDomains = ["bit.ly", "tinyurl.com", "t.co", "short.link"]
            if suspiciousDomains.contains(host) {
                throw SecurityError.suspiciousDomain
            }
        }
    }
    
    // MARK: - Secure Memory Operations
    
    /// Securely clear sensitive data from memory
    public static func secureErase(_ data: inout Data) {
        data.withUnsafeMutableBytes { bytes in
            _ = memset_s(bytes.baseAddress, bytes.count, 0, bytes.count)
        }
    }
    
    /// Securely clear string data
    public static func secureErase(_ string: inout String) {
        var data = Data(string.utf8)
        secureErase(&data)
        string = ""
    }
    
    // MARK: - Timing Attack Protection
    
    /// Constant-time string comparison to prevent timing attacks
    public static func constantTimeCompare(_ a: String, _ b: String) -> Bool {
        let aData = Data(a.utf8)
        let bData = Data(b.utf8)
        return constantTimeCompare(aData, bData)
    }
    
    /// Constant-time data comparison
    public static func constantTimeCompare(_ a: Data, _ b: Data) -> Bool {
        guard a.count == b.count else { return false }
        
        var result: UInt8 = 0
        for i in 0..<a.count {
            result |= a[i] ^ b[i]
        }
        return result == 0
    }
    
    // MARK: - QR Code Integrity
    
    /// Generate integrity hash for QR code data
    public static func generateIntegrityHash(for data: String) -> String {
        if #available(iOS 13.0, macOS 10.15, *) {
            let hash = SHA256.hash(data: Data(data.utf8))
            return hash.compactMap { String(format: "%02x", $0) }.joined()
        } else {
            // Fallback for iOS 12 - use CommonCrypto
            return generateLegacyHash(for: data)
        }
    }
    
    /// Fallback hash generation for iOS 12
    private static func generateLegacyHash(for data: String) -> String {
        // Simple hash implementation for iOS 12 compatibility
        let inputData = Data(data.utf8)
        let hash = inputData.withUnsafeBytes { bytes in
            var digest = [UInt8](repeating: 0, count: 32)
            // Simple XOR-based hash (not cryptographically secure, but functional)
            for (index, byte) in bytes.enumerated() {
                digest[index % 32] ^= byte
            }
            return digest
        }
        return hash.map { String(format: "%02x", $0) }.joined()
    }
    
    /// Verify QR code integrity
    public static func verifyIntegrity(data: String, expectedHash: String) -> Bool {
        let actualHash = generateIntegrityHash(for: data)
        return constantTimeCompare(actualHash, expectedHash)
    }
}

// MARK: - Security Errors

public enum SecurityError: Error, LocalizedError {
    case rateLimitExceeded
    case inputTooLong
    case potentialInjection
    case invalidURL
    case unsafeURLScheme
    case suspiciousDomain
    case integrityCheckFailed
    
    public var errorDescription: String? {
        switch self {
        case .rateLimitExceeded:
            return "Too many operations. Please wait before trying again."
        case .inputTooLong:
            return "Input data exceeds maximum allowed length."
        case .potentialInjection:
            return "Input contains potentially malicious content."
        case .invalidURL:
            return "Invalid URL format detected."
        case .unsafeURLScheme:
            return "URL scheme not allowed for security reasons."
        case .suspiciousDomain:
            return "Domain flagged as potentially suspicious."
        case .integrityCheckFailed:
            return "QR code integrity verification failed."
        }
    }
    
    public var recoverySuggestion: String? {
        switch self {
        case .rateLimitExceeded:
            return "Wait a moment and try again. If this persists, contact support."
        case .inputTooLong:
            return "Reduce the amount of data in the QR code."
        case .potentialInjection:
            return "Remove any script-like content from the input."
        case .invalidURL:
            return "Check the URL format and try again."
        case .unsafeURLScheme:
            return "Use https:// or other standard schemes."
        case .suspiciousDomain:
            return "Use a trusted domain or contact support if this is incorrect."
        case .integrityCheckFailed:
            return "The QR code may have been tampered with. Generate a new one."
        }
    }
} 

File: Android/sample-app/qrcode-sdk/src/main/java/com/qrcodesdk/security/SecurityManager.kt
package com.qrcodesdk.security

import java.net.URL
import java.security.MessageDigest
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicLong

/**
 * Security manager for QR code operations
 */
object SecurityManager {
    
    // MARK: - Rate Limiting
    
    private data class RateLimitInfo(
        val count: AtomicInteger = AtomicInteger(0),
        val lastReset: AtomicLong = AtomicLong(System.currentTimeMillis())
    )
    
    private val operationCounts = ConcurrentHashMap<String, RateLimitInfo>()
    private const val MAX_OPERATIONS_PER_MINUTE = 60
    private const val RATE_LIMIT_WINDOW = 60_000L // 1 minute in milliseconds
    
    /**
     * Check if operation is within rate limits
     */
    fun checkRateLimit(operation: String): Boolean {
        val now = System.currentTimeMillis()
        val info = operationCounts.getOrPut(operation) { RateLimitInfo() }
        
        // Reset counter if window has passed
        if (now - info.lastReset.get() > RATE_LIMIT_WINDOW) {
            info.count.set(1)
            info.lastReset.set(now)
            return true
        }
        
        // Check if within limits
        val currentCount = info.count.get()
        if (currentCount >= MAX_OPERATIONS_PER_MINUTE) {
            return false
        }
        
        // Increment counter
        info.count.incrementAndGet()
        return true
    }
    
    // MARK: - Input Sanitization
    
    /**
     * Sanitize and validate QR code input data
     */
    @Throws(SecurityException::class)
    fun sanitizeQRInput(input: String): String {
        // Remove null bytes and control characters
        val sanitized = input.replace("\u0000", "")
            .replace(Regex("\\p{Cntrl}"), "")
        
        // Validate length
        if (sanitized.length > 4296) { // QR Code max capacity
            throw SecurityException("Input data exceeds maximum allowed length")
        }
        
        // Check for malicious patterns
        validateInputSafety(sanitized)
        
        return sanitized
    }
    
    /**
     * Validate input for potential security threats
     */
    @Throws(SecurityException::class)
    private fun validateInputSafety(input: String) {
        // Check for script injection patterns
        val dangerousPatterns = listOf(
            "<script", "javascript:", "data:", "vbscript:",
            "onload=", "onerror=", "onclick=", "eval(",
            "document.cookie", "window.location"
        )
        
        val lowercaseInput = input.lowercase()
        dangerousPatterns.forEach { pattern ->
            if (lowercaseInput.contains(pattern)) {
                throw SecurityException("Input contains potentially malicious content")
            }
        }
        
        // Validate URL schemes if input looks like URL
        if (input.startsWith("http") || input.startsWith("ftp")) {
            validateURLSafety(input)
        }
    }
    
    /**
     * Validate URL safety for QR codes containing URLs
     */
    @Throws(SecurityException::class)
    private fun validateURLSafety(urlString: String) {
        try {
            val url = URL(urlString)
            
            // Whitelist allowed schemes
            val allowedSchemes = listOf("https", "http", "tel", "mailto", "sms")
            val scheme = url.protocol?.lowercase()
            if (scheme == null || !allowedSchemes.contains(scheme)) {
                throw SecurityException("URL scheme not allowed for security reasons")
            }
            
            // Check for suspicious domains
            url.host?.lowercase()?.let { host ->
                val suspiciousDomains = listOf("bit.ly", "tinyurl.com", "t.co", "short.link")
                if (suspiciousDomains.contains(host)) {
                    throw SecurityException("Domain flagged as potentially suspicious")
                }
            }
        } catch (e: Exception) {
            if (e is SecurityException) throw e
            throw SecurityException("Invalid URL format detected")
        }
    }
    
    // MARK: - Secure Memory Operations
    
    /**
     * Securely clear sensitive data from memory
     */
    fun secureErase(data: ByteArray) {
        data.fill(0)
    }
    
    /**
     * Securely clear string data
     */
    fun secureErase(string: StringBuilder) {
        repeat(string.length) { string.setCharAt(it, '\u0000') }
        string.clear()
    }
    
    // MARK: - Timing Attack Protection
    
    /**
     * Constant-time string comparison to prevent timing attacks
     */
    fun constantTimeCompare(a: String, b: String): Boolean {
        val aBytes = a.toByteArray()
        val bBytes = b.toByteArray()
        return constantTimeCompare(aBytes, bBytes)
    }
    
    /**
     * Constant-time byte array comparison
     */
    fun constantTimeCompare(a: ByteArray, b: ByteArray): Boolean {
        if (a.size != b.size) return false
        
        var result = 0
        for (i in a.indices) {
            result = result or (a[i].toInt() xor b[i].toInt())
        }
        return result == 0
    }
    
    // MARK: - QR Code Integrity
    
    /**
     * Generate integrity hash for QR code data
     */
    fun generateIntegrityHash(data: String): String {
        val digest = MessageDigest.getInstance("SHA-256")
        val hash = digest.digest(data.toByteArray())
        return hash.joinToString("") { "%02x".format(it) }
    }
    
    /**
     * Verify QR code integrity
     */
    fun verifyIntegrity(data: String, expectedHash: String): Boolean {
        val actualHash = generateIntegrityHash(data)
        return constantTimeCompare(actualHash, expectedHash)
    }
}

/**
 * Security-related exceptions
 */
sealed class SecurityException(message: String, val recoverySuggestion: String? = null) : Exception(message) {
    
    class RateLimitExceeded : SecurityException(
        "Too many operations. Please wait before trying again.",
        "Wait a moment and try again. If this persists, contact support."
    )
    
    class InputTooLong : SecurityException(
        "Input data exceeds maximum allowed length.",
        "Reduce the amount of data in the QR code."
    )
    
    class PotentialInjection : SecurityException(
        "Input contains potentially malicious content.",
        "Remove any script-like content from the input."
    )
    
    class InvalidURL : SecurityException(
        "Invalid URL format detected.",
        "Check the URL format and try again."
    )
    
    class UnsafeURLScheme : SecurityException(
        "URL scheme not allowed for security reasons.",
        "Use https:// or other standard schemes."
    )
    
    class SuspiciousDomain : SecurityException(
        "Domain flagged as potentially suspicious.",
        "Use a trusted domain or contact support if this is incorrect."
    )
    
    class IntegrityCheckFailed : SecurityException(
        "QR code integrity verification failed.",
        "The QR code may have been tampered with. Generate a new one."
    )
} 